beer-garden/functionary
# #23 - Roles and Permissions

Closes #23 

This PR does the following:

* Replaces the `TeamMember` model with two separate models: `TeamUserRole` and `EnvironmentUserRole`.
* Establishes an initial set of Roles and their corresponding permissions, which are just based on CRUD operations for each of the models for now.
* Implements a custom auth backend to allow checking if a user has a given permission for a Team or Environment, based on their assigned `TeamUserRoles` and `EnvironmentUserRoles`.

Note that this PR does not yet implement permissions checks into any of the views.  I avoided doing this for now as to not make this PR grow excessively large.

## Testing Instructions
To see this in action, first user the django admin pages to do the following:
* Create a non-superuser account
* Create a Team and any number of environments.
* Create TeamUserRole and EnvironmentUserRole entries for the user.

Now launch a django shell and do something like the following:
```python
from core.auth import Permissions
from core.models import User, Environment, Team

# In this example, snake has the READ_ONLY role for the foxhound default environment
snake = User.objects.get(username="snake")
team = Team.objects.get(name="foxhound")
environment = Environment.objects.get(team=team, name="default")

snake.team_permissions(team)
# set()

snake.environment_permissions(environment)
"""
{'environment:read',
 'function:read',
 'package:read',
 'team:read',
 'user:read',
 'userrole:read'}
"""

snake.has_perm(Permissions.PACKAGE_READ, environment)
# True

snake.has_perm(Permissions.PACKAGE_READ, team)
# False

snake.has_perm(Permissions.PACKAGE_CREATE, environment)
# False
```

If we granted snake the ADMIN role at the foxhound team level, we would see this:

```python
snake.team_permissions(team)
"""
{'environment:create',
 'environment:delete',
 'environment:read',
 'environment:update',
 'function:create',
 'function:delete',
 'function:read',
 'function:update',
 'package:create',
 'package:delete',
 'package:read',
 'package:update',
 'team:create',
 'team:delete',
 'team:read',
 'team:update',
 'user:create',
 'user:delete',
 'user:read',
 'user:update',
 'userrole:create',
 'userrole:delete',
 'userrole:read',
 'userrole:update'}
"""

snake.environment_permissions(environment)
"""
{'environment:read',
 'function:read',
 'package:read',
 'team:read',
 'user:read',
 'userrole:read'}
"""

snake.environment_permissions(environment, inherited=True)
"""
{'environment:create',
 'environment:delete',
 'environment:read',
 'environment:update',
 'function:create',
 'function:delete',
 'function:read',
 'function:update',
 'package:create',
 'package:delete',
 'package:read',
 'package:update',
 'team:create',
 'team:delete',
 'team:read',
 'team:update',
 'user:create',
 'user:delete',
 'user:read',
 'user:update',
 'userrole:create',
 'userrole:delete',
 'userrole:read',
 'userrole:update'}
"""

snake.has_perm(Permissions.PACKAGE_CREATE, team)
# True

snake.has_perm(Permissions.PACKAGE_CREATE, environment)
# True
```
